### PR TITLE
Add ValueTask and ConfigureAwait details to "info & advice"

### LINKED ISSUE #18225: Additional Advice: ConfigureAwait and ValueTask
Great article!

I found the last section (Important Info and Advice) particularly useful. I was wondering if there's a place for adding the case for ConfigureAwait and ValueTask (separately) as well.

From my experience, although async/await and tasks are powerful abstractions, and developers adopt them fast, some threat them as "magic" and sooner or later find themselves in a bind. I think your opinionated advice helps put the spotlight on important pitfalls and guide developers in the right direction for further inquiry. I wish I would have stumbled across this article when I started writing asynchronous code. Good job!

---
#### פרטי המסמך

⚠ *אין לערוך מקטע זה. פעולה זו נדרשת עבור קישור לנושאים ב- docs.microsoft.com ← GitHub.*

* ID: 07ed82a3-99c1-0893-4e81-d095b4757ecc
* Version Independent ID: 3765200b-43e5-441c-b290-709c34dd2f6d
* Content: [Asynchronous programming - C#](https://docs.microsoft.com/he-il/dotnet/csharp/async#feedback)
* Content Source: [docs/csharp/async.md](https://github.com/dotnet/docs/blob/master/docs/csharp/async.md)
* Product: **dotnet-csharp**
* Technology: **csharp-async**
* GitHub Login: @cartermp
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- General clean up of article
    - Acrolinx
    - Sentence case headings
    - Add some `xref`'s
    - Add link to `ConfigureAwait` FAQ
    - Add code fence around missing segments
    - Correctly tab C#
    - Replace "  " double space after sentences, with " " single space
    - Format table
- Added `ValueTask` and `ConfigureAwait` in "info & advice"

Fixes #18225 

#### Preview

✔️ [Asynchronous programming](https://review.docs.microsoft.com/en-us/dotnet/csharp/async?branch=pr-en-us-18527)

### CODE REVIEW

File: docs/csharp/async.md [hunk 1]
```diff
@@ -2,129 +2,128 @@
...
 
-If the work you have is **I/O-bound**, use `async` and `await` *without* `Task.Run`.  You *should not* use the Task Parallel Library.  The reason for this is outlined in the [Async in Depth article](../standard/async-in-depth.md).
+If the work you have is **I/O-bound**, use `async` and `await` *without* `Task.Run`. You *should not* use the Task Parallel Library. The reason for this is outlined in the [Async in Depth article](../standard/async-in-depth.md).
 
-If the work you have is **CPU-bound** and you care about responsiveness, use `async` and `await` but spawn the work off on another thread *with* `Task.Run`.  If the work is appropriate for concurrency and parallelism, you should also consider using the [Task Parallel Library](../standard/parallel-programming/task-parallel-library-tpl.md).
+If the work you have is **CPU-bound** and you care about responsiveness, use `async` and `await` but spawn off the work on another thread *with* `Task.Run`. If the work is appropriate for concurrency and parallelism, you should also consider using the [Task Parallel Library](../standard/parallel-programming/task-parallel-library-tpl.md).
```
Comment (gewarren) [inline docs/csharp/async.md:102]:
```suggestion
If the work you have is **CPU-bound** and you care about responsiveness, use `async` and `await`, but spawn off the work on another thread *with* `Task.Run`. If the work is appropriate for concurrency and parallelism, also consider using the [Task Parallel Library](../standard/parallel-programming/task-parallel-library-tpl.md).
```

File: docs/csharp/async.md [hunk 2]
```diff
@@ -2,129 +2,128 @@
...
+Additionally, you should always measure the execution of your code. For example, you may find yourself in a situation where your CPU-bound work is not costly enough compared with the overhead of context switches when multithreading. Every choice has its tradeoff, and you should pick the correct tradeoff for your situation.
 
-## More Examples
+## More examples
 
-The following examples demonstrate various ways you can write async code in C#.  They cover a few different scenarios you may come across.
+The following examples demonstrate various ways you can write async code in C#. They cover a few different scenarios you may come across.
 
-### Extracting Data from a Network
+### Extracting data from a network
```
Comment (gewarren) [inline docs/csharp/async.md:110]:
```suggestion
### Extract data from a network
```

File: docs/csharp/async.md [hunk 3]
```diff
@@ -2,129 +2,128 @@
...
 
-This snippet downloads the HTML from the homepage at [www.dotnetfoundation.org](https://www.dotnetfoundation.org) and counts the number of times the string ".NET" occurs in the HTML.  It uses ASP.NET MVC to define a web controller method which performs this task, returning the number.
+This snippet downloads the HTML from the homepage at <https://dotnetfoundation.org> and counts the number of times the string ".NET" occurs in the HTML. It uses ASP.NET to define a Web API controller method, which performs this task, returning the number.
```
Comment (gewarren) [inline docs/csharp/async.md:112]:
```suggestion
This snippet downloads the HTML from the homepage at <https://dotnetfoundation.org> and counts the number of times the string ".NET" occurs in the HTML. It uses ASP.NET to define a Web API controller method, which performs this task and returns the number.
```

Comment (gewarren) [reply docs/csharp/async.md:112]: I haven't seen angle brackets used for links like that. What is the purpose?

File: docs/csharp/async.md [hunk 4]
```diff
@@ -160,9 +159,9 @@ private async void SeeTheDotNets_Click(object sender, RoutedEventArgs e)
 }
 ```
 
-### Waiting for Multiple Tasks to Complete
+### Waiting for multiple tasks to complete
```
Comment (gewarren) [inline docs/csharp/async.md:162]:
```suggestion
### Wait for multiple tasks to complete
```

File: docs/csharp/async.md [hunk 5]
```diff
@@ -160,9 +159,9 @@ private async void SeeTheDotNets_Click(object sender, RoutedEventArgs e)
...
 
-You may find yourself in a situation where you need to retrieve multiple pieces of data concurrently.  The `Task` API contains two methods, `Task.WhenAll` and `Task.WhenAny` which allow you to write asynchronous code which performs a non-blocking wait on multiple background jobs.
+You may find yourself in a situation where you need to retrieve multiple pieces of data concurrently. The `Task` API contains two methods, <xref:System.Threading.Tasks.Task.WhenAll%2A?displayProperty=nameWithType> and <xref:System.Threading.Tasks.Task.WhenAny%2A?displayProperty=nameWithType> that allow you to write asynchronous code, which performs a non-blocking wait on multiple background jobs.
```
Comment (gewarren) [inline docs/csharp/async.md:164]:
```suggestion
You may find yourself in a situation where you need to retrieve multiple pieces of data concurrently. The `Task` API contains two methods, <xref:System.Threading.Tasks.Task.WhenAll%2A?displayProperty=nameWithType> and <xref:System.Threading.Tasks.Task.WhenAny%2A?displayProperty=nameWithType>, that allow you to write asynchronous code that performs a non-blocking wait on multiple background jobs.
```

File: docs/csharp/async.md [hunk 6]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
-## Important Info and Advice
+## Important info and advice
 
-Although async programming is relatively straightforward, there are some details to keep in mind which can prevent unexpected behavior.
+With async programming there are some details to keep in mind which can prevent unexpected behavior.
 
 * `async` **methods need to have an** `await` **keyword in their body or they will never yield!**
 
-This is important to keep in mind.  If `await` is not used in the body of an `async` method, the C# compiler will generate a warning, but the code will compile and run as if it were a normal method.  Note that this would also be incredibly inefficient, as the state machine generated by the C# compiler for the async method would not be accomplishing anything.
+This is important to keep in mind. If `await` is not used in the body of an `async` method, the C# compiler will generate a warning, but the code will compile and run as if it were a normal method. Note that this would also be incredibly inefficient, as the state machine generated by the C# compiler for the async method would not be accomplishing anything.
```
Comment (gewarren) [inline docs/csharp/async.md:215]:
```suggestion
This is important to keep in mind. If `await` is not used in the body of an `async` method, the C# compiler generates a warning, but the code compiles and runs as if it were a normal method. This is incredibly inefficient, as the state machine generated by the C# compiler for the async method is not accomplishing anything.
```

File: docs/csharp/async.md [hunk 7]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
 
 * **You should add "Async" as the suffix of every async method name you write.**
 
-This is the convention used in .NET to more-easily differentiate synchronous and asynchronous methods. Note that certain methods which aren’t explicitly called by your code (such as event handlers or web controller methods) don’t necessarily apply. Because these are not explicitly called by your code, being explicit about their naming isn’t as important.
+This is the convention used in .NET to more-easily differentiate synchronous and asynchronous methods. Certain methods that aren't explicitly called by your code (such as event handlers or web controller methods) don't necessarily apply. Because they are not explicitly called by your code, being explicit about their naming isn't as important.
```
Comment (gewarren) [inline docs/csharp/async.md:219]:
```suggestion
This is the convention used in .NET to more easily differentiate synchronous and asynchronous methods. Certain methods that aren't explicitly called by your code (such as event handlers or web controller methods) don't necessarily apply. Because they are not explicitly called by your code, being explicit about their naming isn't as important.
```

File: docs/csharp/async.md [hunk 8]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
 
-* `async void` **should only be used for event handlers.**
+* `async void` **is only to be used for event handlers.**
```
Comment (gewarren) [inline docs/csharp/async.md:221]: Why did you change the wording here? I think should is clearer.

File: docs/csharp/async.md [hunk 9]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
+Lambda expressions in LINQ use deferred execution, meaning code could end up executing at a time when you're not expecting it to. The introduction of blocking tasks into this can easily result in a deadlock if not written correctly. Additionally, the nesting of asynchronous code like this can also make it more difficult to reason about the execution of the code. Async and LINQ are powerful, but should be used together as carefully and clearly as possible.
 
 * **Write code that awaits Tasks in a non-blocking manner**
 
-Blocking the current thread as a means to wait for a Task to complete can result in deadlocks and blocked context threads, and can require significantly more complex error-handling. The following table provides guidance on how to deal with waiting for Tasks in a non-blocking way:
+Blocking the current thread as a means to wait for a `Task` to complete can result in deadlocks and blocked context threads, and can require more complex error-handling. The following table provides guidance on how to deal with waiting for Tasks in a non-blocking way:
```
Comment (gewarren) [inline docs/csharp/async.md:235]:
```suggestion
Blocking the current thread as a means to wait for a `Task` to complete can result in deadlocks and blocked context threads, and can require more complex error-handling. The following table provides guidance on how to deal with waiting for tasks in a non-blocking way:
```

File: docs/csharp/async.md [hunk 10]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
+
+| Use this...          | Instead of this...           | When wishing to do this                    |
```
Comment (gewarren) [inline docs/csharp/async.md:237]:
```suggestion
| Use this...          | Instead of this...           | When wishing to do this...                    |
```

Just to be consistent.

File: docs/csharp/async.md [hunk 11]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
+|----------------------|------------------------------|--------------------------------------------|
+| `await`              | `Task.Wait` or `Task.Result` | Retrieving the result of a background task |
+| `await Task.WhenAny` | `Task.WaitAny`               | Waiting for any task to complete           |
+| `await Task.WhenAll` | `Task.WaitAll`               | Waiting for all tasks to complete          |
+| `await Task.Delay`   | `Thread.Sleep`               | Waiting for a period of time               |
+
+* **Consider using** `ValueTask` **where possible**
+
+Returning a `Task` object from async methods can introduce performance bottlenecks in certain paths. `Task` is a reference type, so using it means allocating an object. In cases where a method declared with the `async` modifier returns a cached result, or completes synchronously, the extra allocations can become a significant time cost in performance critical sections of code. It can become costly if those allocations occur in tight loops. For more information, see [generalized async return types](whats-new/csharp-7.md#generalized-async-return-types).
```
Comment (gewarren) [inline docs/csharp/async.md:246]:
```suggestion
Returning a `Task` object from async methods can introduce performance bottlenecks in certain paths. `Task` is a reference type, so using it means allocating an object. In cases where a method declared with the `async` modifier returns a cached result or completes synchronously, the extra allocations can become a significant time cost in performance critical sections of code. It can become costly if those allocations occur in tight loops. For more information, see [generalized async return types](whats-new/csharp-7.md#generalized-async-return-types).
```

File: docs/csharp/async.md [hunk 12]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
+
+* **Consider using** `ConfigureAwait(false)`
 
-| Use this... | Instead of this... | When wishing to do this |
-| --- | --- | --- |
-| `await` | `Task.Wait` or `Task.Result` | Retrieving the result of a background task |
-| `await Task.WhenAny` | `Task.WaitAny` | Waiting for any task to complete |
-| `await Task.WhenAll` | `Task.WaitAll` | Waiting for all tasks to complete |
-| `await Task.Delay` | `Thread.Sleep` | Waiting for a period of time |
+A common question is, "when should I use the <xref:System.Threading.Tasks.Task.ConfigureAwait(System.Boolean)?displayProperty=nameWithType> method?". The method allows for a `Task` instance to configure its awaiter. This is an important consideration and setting incorrectly could potentially have performance implications and even deadlocks. For more information on `ConfigureAwait`, see the [ConfigureAwait FAQ](https://devblogs.microsoft.com/dotnet/configureawait-faq).
```
Comment (gewarren) [inline docs/csharp/async.md:250]:
```suggestion
A common question is, "when should I use the <xref:System.Threading.Tasks.Task.ConfigureAwait(System.Boolean)?displayProperty=nameWithType> method?". The method allows for a `Task` instance to configure its awaiter. This is an important consideration and setting it incorrectly could potentially have performance implications and even deadlocks. For more information on `ConfigureAwait`, see the [ConfigureAwait FAQ](https://devblogs.microsoft.com/dotnet/configureawait-faq).
```

**Review by gewarren — APPROVED** [review]
Left a number of suggestions.

File: docs/csharp/async.md (hunk 3 above)
Comment (IEvangelist) [reply docs/csharp/async.md:112]: It treats it as a link, simplifies it. They were wrong, since the `www.` isn't needed with this site and redirects anyways. So rather than updating it, I simplified it.

File: docs/csharp/async.md (hunk 8 above)
Comment (IEvangelist) [reply docs/csharp/async.md:221]: Acrolinx didn't like "should", I'll change it back.

File: docs/csharp/async.md [hunk 13]
```diff
@@ -2,129 +2,128 @@
...
 
-The first two examples of this guide showed how you can use `async` and `await` for I/O-bound and CPU-bound work.  It's key that you can identify when a job you need to do is I/O-bound or CPU-bound, because it can greatly affect the performance of your code and could potentially lead to misusing certain constructs.
+The first two examples of this guide showed how you can use `async` and `await` for I/O-bound and CPU-bound work. It's key that you can identify when a job you need to do is I/O-bound or CPU-bound, because it can greatly affect the performance of your code and could potentially lead to misusing certain constructs.
 
 Here are two questions you should ask before you write any code:
 
 1. Will your code be "waiting" for something, such as data from a database?
 
-    If your answer is "yes", then your work is **I/O-bound**.
+  If your answer is "yes", then your work is **I/O-bound**.
```
Comment (gewarren) [inline docs/csharp/async.md:94]:
```suggestion
   If your answer is "yes", then your work is **I/O-bound**.
```

Without proper indent, the second item in the list is also numbered as 1.

File: docs/csharp/async.md [hunk 14]
```diff
@@ -2,129 +2,128 @@
...
 
-2. Will your code be performing a very expensive computation?
+1. Will your code be performing an expensive computation?
 
-    If you answered "yes", then your work is **CPU-bound**.
+  If you answered "yes", then your work is **CPU-bound**.
```
Comment (gewarren) [inline docs/csharp/async.md:98]:
```suggestion
   If you answered "yes", then your work is **CPU-bound**.
```

File: docs/csharp/async.md [hunk 15]
```diff
@@ -206,46 +204,54 @@ public static async Task<User[]> GetUsersAsync(IEnumerable<int> userIds)
...
-## Important Info and Advice
+## Important info and advice
 
-Although async programming is relatively straightforward, there are some details to keep in mind which can prevent unexpected behavior.
+With async programming there are some details to keep in mind which can prevent unexpected behavior.
 
 * `async` **methods need to have an** `await` **keyword in their body or they will never yield!**
 
-This is important to keep in mind.  If `await` is not used in the body of an `async` method, the C# compiler will generate a warning, but the code will compile and run as if it were a normal method.  Note that this would also be incredibly inefficient, as the state machine generated by the C# compiler for the async method would not be accomplishing anything.
+This is important to keep in mind. If `await` is not used in the body of an `async` method, the C# compiler generates a warning, but the code compiles and runs as if it were a normal method. This is incredibly inefficient, as the state machine generated by the C# compiler for the async method is not accomplishing anything.
```
Comment (gewarren) [inline docs/csharp/async.md:215]:
```suggestion
  This is important to keep in mind. If `await` is not used in the body of an `async` method, the C# compiler generates a warning, but the code compiles and runs as if it were a normal method. This is incredibly inefficient, as the state machine generated by the C# compiler for the async method is not accomplishing anything.
```

I would apply the same indentation to all the bullet points so the text aligns better.

**Review by gewarren — COMMENTED** [review]
Noticed a couple more things on the preview.
